nucypher/network/sensors.py

```diff
@@ -0,0 +1,196 @@
+        if not result:
+            for time, reason in self.__excuses.items():
+                self.log.info(f'[{time}] - {reason["error"]}')
+        return result
+
+    @property
+    def running(self) -> bool:
+        return self.__task.running
+
+    def start(self, now: bool = False):
+        if not self.running:
+            self._start_time = maya.now()
+            d = self.__task.start(interval=self.FAST_INTERVAL, now=now)
+            d.addErrback(self.handle_measurement_errors)
+
+    def stop(self) -> None:
+        if self.running:
+            self.__task.stop()
+
+    def maintain(self) -> None:
+        self.log.debug(f"Starting new sensor maintenance round")
+        known_nodes_is_smaller_than_sample_size = len(self._ursula.known_nodes) < self.SAMPLE_SIZE
+
+        # If there are no known nodes or too few known nodes, skip this round...
+        # ... but not for longer than the maximum allotted alone time
+        if known_nodes_is_smaller_than_sample_size:
+            if not self._ursula.lonely and self.enforce_loneliness:
+                now = maya.now().epoch
+                delta = now - self._start_time.epoch
+                if delta >= self.MAXIMUM_ALONE_TIME:
+                    self.severe_warning()
+                    reason = self.Solitary if not self._ursula.known_nodes else self.Lonely
+                    self.shutdown_everything(reason=reason)
+            return
+
+        if self.__task.interval == self.FAST_INTERVAL:
+            now = maya.now().epoch
+            delta = now - self._start_time.epoch
+            if delta >= self.SEEDING_DURATION:
+                # Slow down
+                self.__task.interval = self.SLOW_INTERVAL
+                return
+
+        # All systems go
+        self.measure()
+        delta = maya.now() - self._start_time
+        self.log.info(f"Current availability score is {self.score} measured since {delta}")
+        self.issue_warnings()
+
+    def issue_warnings(self, cascade: bool = True) -> None:
+        warnings = sorted(self.warnings.items(), key=lambda t: t[0])
+        for threshold, action in warnings:
+            if self.score <= threshold:
+                action()
+                if not cascade:
+                    # Exit after the first active warning is issued
+                    return
+
+    def sample(self, quantity: int) -> list:
+        population = tuple(self._ursula.known_nodes._nodes.values())
+        ursulas = random.sample(population=population, k=quantity)
+        return ursulas
+
+    @property
+    def score(self) -> float:
+        return self.__score
+
+    def record(self, result: bool = None, reason: dict = None) -> None:
+        """Score the result and cache it."""
+        if (not result) and reason:
+            self.__excuses[maya.now().epoch] = reason
+        if result is None:
+            return  # Actually nevermind, dont score this one...
+        score = int(result) + self.CHARGE_RATE * self.__score
```

Comment:
@KPrasch So if result is 'false' then `score = 0 + 0.9 * <previous_score>" - so good.

However, do we ever actually hit 0 for a shutdown, or we just get very close to it over time?